src/Nix/Reduce.hs

```diff
@@ -155,10 +165,41 @@ reduce (NBinary_ bann op larg rarg) = do
             return $ Fix (NConstant_ ann (NInt (x + y)))
         _ -> pure $ Fix $ NBinary_ bann op lval rval
 
--- reduce (NSelect aset attr alt) = do
+-- | Reduce a select on a Set by substituing the set to the selected value.
+--
+-- Before applying this reduction, we need to ensure that:
+--
+--   1. The selected expr is indeed a set.
+--   2. The selection AttrPath is a list of StaticKeys.
+--   3. The selected AttrPath exists in the set.
+reduce base@(NSelect_ _ _ attrs _)
+    | sAttrPath $ NE.toList attrs = do
+      (NSelect_ _ aset attrs _) <- sequence base
+      inspectSet (unFix aset) attrs
+    | otherwise = sId
+  where
+    sId = Fix <$> sequence base
+    -- The selection AttrPath is composed of StaticKeys.
+    sAttrPath (StaticKey _:xs) = sAttrPath xs
+    sAttrPath []               = True
+    sAttrPath _                = False
+    -- Find appropriate bind in set's binds.
+    findBind [] _                = Nothing
+    findBind (x:xs) attrs@(a:|_) = case x of
+        n@(NamedVar (a':|_) _ _) | a' == a -> Just n
+        _                                  -> findBind xs attrs
+    -- Follow the attrpath recursively in sets.
+    inspectSet (NSet_ _ binds) attrs = case findBind binds attrs of
+       Just (NamedVar _ e _) -> case NE.uncons attrs of 
+               (_,Just attrs) -> inspectSet (unFix e) attrs
+               _              -> pure e
+       _ -> sId
+    inspectSet _ _ = sId
```

Comment:
All looks good to me, and I like that you've added tests. I'd say this ready for inclusion.